File: src/Commands/CommandsFactory.ts
import { Header } from "./All/Header";
import { PageRef } from "./All/PageRef";
import { Remember } from "./All/Remember";
import { Tag } from "./All/Tag";
import { AutoReveal } from "./All/AutoReveal";
import { Alias } from "./All/Alias";
import { Recap } from "./All/Recap";
import { Example } from "./All/Example";
import { RecallData } from "./All/RecallData";
import { LineDescriptor, IStudySet } from "@/FlashcardParser/Types/Types";

export class CommandsFactory {
    static Make(
        lineDescriptor: LineDescriptor,
        studySet: IStudySet,
        commandNameOrShortcut: string,
        commandArgument: string | null,
    ) {
        if (commandNameOrShortcut[0] === '\\') {
            commandNameOrShortcut = commandNameOrShortcut.slice(1);
        }

        let p = commandNameOrShortcut.indexOf('(');
        if (p !== -1) {
            commandNameOrShortcut = commandNameOrShortcut.slice(0, p);
        }

        switch (commandNameOrShortcut) {
            case PageRef.symbol:
            case 'page':
                return new PageRef(commandArgument);
            case Remember.symbol:
            case 'remember':
                return new Remember(commandArgument);
            case Tag.symbol:
            case 'tag':
                return new Tag(commandArgument || "");
            case AutoReveal.symbol:
            case 'auto_reveal':
                return new AutoReveal();
            case Alias.symbol:
            case 'alias':
                const alias = new Alias(commandArgument, lineDescriptor, studySet);
                return alias.isValid ? alias : null
            case Header.symbol:
            case 'header':
                return new Header(lineDescriptor, studySet, commandArgument || 1);
            case RecallData.symbol:
            case 'recall_data':
                return new RecallData(lineDescriptor, studySet, commandArgument || '');
            case 'recap':
            case 'Recap':
                return new Recap(lineDescriptor, commandArgument);
            case 'example':
            case 'Example':
                return new Example(lineDescriptor, commandArgument);
            default:
                console.log(`[MakeCommand] Unrecognized command ${commandNameOrShortcut}`)
                return null;
        }
    }
}

File: service/flashcard_parser/commands/commands_factory.py
from ..types import LineDescriptor, StudySet
from .page_ref import PageRef
from .remember import Remember
from .tag import Tag
from .auto_reveal import AutoReveal
from .alias_command import Alias as AliasCommand
from .header import Header
from .recall_data import RecallData
from .recap import Recap
from .example import Example
from .note import Note


class CommandsFactory:
    @staticmethod
    def make(line_descriptor: LineDescriptor, study_set: StudySet, command_name_or_shortcut: str, command_argument: str | None):
        if command_name_or_shortcut.startswith("\\"):
            command_name_or_shortcut = command_name_or_shortcut[1:]
        p = command_name_or_shortcut.find("(")
        if p != -1:
            command_name_or_shortcut = command_name_or_shortcut[:p]
        if command_name_or_shortcut in (PageRef.symbol, "page"):
            return PageRef(command_argument or "0")
        if command_name_or_shortcut in (Remember.symbol, "remember"):
            return Remember(command_argument or "")
        if command_name_or_shortcut in (Tag.symbol, "tag"):
            return Tag(command_argument or "")
        if command_name_or_shortcut in (AutoReveal.symbol, "auto_reveal"):
            return AutoReveal()
        if command_name_or_shortcut in (AliasCommand.symbol, "alias"):
            alias = AliasCommand(command_argument or "", line_descriptor, study_set)
            return alias if alias.is_valid else None
        if command_name_or_shortcut in (Header.symbol, "header"):
            return Header(line_descriptor, study_set, command_argument or 1)
        if command_name_or_shortcut in (RecallData.symbol, "recall_data"):
            return RecallData(line_descriptor, study_set, command_argument or "")
        if command_name_or_shortcut.lower() == "recap":
            return Recap(line_descriptor, command_argument)
        if command_name_or_shortcut.lower() == "example":
            return Example(line_descriptor, command_argument)
        if command_name_or_shortcut.lower() == "note":
            return Note(command_argument or "")
        print(f"[MakeCommand] Unrecognized command {command_name_or_shortcut}")
        return None
